bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/TestVerificationSampling.java

```diff
@@ -0,0 +1,241 @@
+  }
+
+  @Test
+  public void isGetListSampled() throws IOException {
+    doReturn(new Result[] {createResult("k")})
+        .when(primaryTable)
+        .get(ArgumentMatchers.<Get>anyList());
+
+    withSamplingEnabled(false);
+    mirroringTable.get(gets);
+    verify(readSampler, times(1)).shouldNextReadOperationBeSampled();
+    verify(primaryTable, times(1)).get(gets);
+    verify(secondaryTable, never()).get(gets);
+
+    withSamplingEnabled(true);
+    mirroringTable.get(gets);
+    executorServiceRule.waitForExecutor();
+    verify(readSampler, times(2)).shouldNextReadOperationBeSampled();
+    verify(primaryTable, times(2)).get(gets);
+    verify(secondaryTable, times(1)).get(gets);
+  }
+
+  @Test
+  public void isExistsSampled() throws IOException {
+    withSamplingEnabled(false);
+    mirroringTable.exists(get);
+    verify(readSampler, times(1)).shouldNextReadOperationBeSampled();
+    verify(primaryTable, times(1)).exists(get);
+    verify(secondaryTable, never()).exists(get);
+
+    withSamplingEnabled(true);
+    mirroringTable.exists(get);
+    executorServiceRule.waitForExecutor();
+    verify(readSampler, times(2)).shouldNextReadOperationBeSampled();
+    verify(primaryTable, times(2)).exists(get);
+    verify(secondaryTable, times(1)).exists(get);
+  }
+
+  @Test
+  public void isExistsAllSampled() throws IOException {
+    doReturn(new boolean[] {true}).when(primaryTable).existsAll(ArgumentMatchers.<Get>anyList());
+
+    withSamplingEnabled(false);
+    mirroringTable.existsAll(gets);
+    verify(readSampler, times(1)).shouldNextReadOperationBeSampled();
+    verify(primaryTable, times(1)).existsAll(gets);
+    verify(secondaryTable, never()).existsAll(gets);
+
+    withSamplingEnabled(true);
+    mirroringTable.existsAll(gets);
+    executorServiceRule.waitForExecutor();
+    verify(readSampler, times(2)).shouldNextReadOperationBeSampled();
+    verify(primaryTable, times(2)).existsAll(gets);
+    verify(secondaryTable, times(1)).existsAll(gets);
+  }
+
+  @Test
+  public void isBatchSampled() throws IOException, InterruptedException {
+    Put put = createPut("test", "test", "test", "test");
+    List<? extends Row> ops = ImmutableList.of(get, put);
+
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocationOnMock) throws Throwable {
+                Object[] args = invocationOnMock.getArguments();
+                Object[] result = (Object[]) args[1];
+                result[0] = Result.create(new Cell[0]);
+                result[1] = Result.create(new Cell[0]);
+                return null;
+              }
+            })
+        .when(primaryTable)
+        .batch(eq(ops), any(Object[].class));
+
+    withSamplingEnabled(false);
+    mirroringTable.batch(ops);
+    verify(readSampler, times(1)).shouldNextReadOperationBeSampled();
+    verify(primaryTable, times(1)).batch(eq(ops), any(Object[].class));
```

Comment:
What about verifying that secondary is never called?

Reply:
The secondary should be called with only the `put` object, the `get` object should be omitted. Tests were improved.